x-pack/plugins/ml/public/application/components/field_stats_flyout/field_stats_flyout_provider.tsx

```diff
@@ -27,6 +47,68 @@ export const FieldStatsFlyoutProvider: FC<{
     () => setFieldStatsIsFlyoutVisible(!isFieldStatsFlyoutVisible),
     [isFieldStatsFlyoutVisible]
   );
+  const [manager] = useState(new PopulatedFieldsCacheManager());
+  const [populatedFields$] = useState(new BehaviorSubject<Set<string>>(new Set()));
+
+  useEffect(() => {
+    if (disablePopulatedFields) return;
+    const abortController = new AbortController();
+
+    const queryAndRunTimeMappings = getMergedSampleDocsForPopulatedFieldsQuery({
+      searchQuery: dslQuery,
+      runtimeFields: dataView.getRuntimeMappings(),
+      datetimeField: dataView.getTimeField()?.name,
+      timeRange: timeRangeMs,
+    });
+    const indexPattern = dataView.getIndexPattern();
+    const esSearchRequestParams = {
+      index: indexPattern,
+      body: {
+        fields: ['*'],
+        _source: false,
+        ...queryAndRunTimeMappings,
+        size: 1000,
+      },
+    };
+    const cacheKey = stringHash(JSON.stringify(esSearchRequestParams)).toString();
+
+    const fetchSampleDocuments = async function () {
+      try {
+        const resp = await lastValueFrom(
+          search.search(
+            {
+              params: esSearchRequestParams,
+            },
+            { abortSignal: abortController.signal }
+          )
+        );
+
+        const docs = resp.rawResponse.hits.hits.map((d) => getProcessedFields(d.fields ?? {}));
+
+        // Get all field names for each returned doc and flatten it
+        // to a list of unique field names used across all docs.
+        // const allDataViewFields = getFieldsFromKibanaIndexPattern(dataView);
+        const fieldsWithData = new Set(docs.map(Object.keys).flat(1));
+        manager.set(cacheKey, fieldsWithData);
+        populatedFields$.next(fieldsWithData);
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error(
+          `An error occurred fetching sample documents to determine populated field stats.
+          \nQuery:\n${JSON.stringify(esSearchRequestParams)}
+          \nError:${e}`
+        );
+      }
+    };
+
+    const cachedResult = manager.get(cacheKey);
+    if (cachedResult) {
+      return cachedResult;
+    } else {
+      fetchSampleDocuments();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [JSON.stringify({ query: dslQuery, dataViewId: dataView.id })]);
```

Comment:
did you consider memoizing `dslQuery` to avoid using `JSON.stringify`?

Reply:
This was actually missing another dependency, which is also another object. I think it's fine here for now.